add sameness group to service resolver, update manifests

malizz opened the pull request:
Changes proposed in this PR:
- draft PR for adding sameness group to service resolver
- added validation for service resolver struct, its subsets, failover and redirest fields
- added and updated tests 

How I've tested this PR: will test locally following test instructions [here](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#testing-a-new-crd), will pick the test up tomorrow as consul build on main crashes atm

How I expect reviewers to test this PR:


Checklist:
- [x] Tests added
- NA CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

wilkermichael commented:
@malizz Just so you know, I don't believe slack gives notifications when you move this from `draft` to `ready for review`. A bunch of us are using the slack integration, so just a heads up. I only seem to get notified on the initial review request. So just tag me in a message here when this is ready for review.

wilkermichael commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -374,9 +446,76 @@ func (in *ServiceResolverRedirect) toConsul() *capi.ServiceResolverRedirect {
 	}
 }
 
+func (in *ServiceResolverRedirect) validate(path *field.Path, consulMeta common.ConsulMeta) field.ErrorList {
```
If possible can you add unit test coverage for this validation?

wilkermichael commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -374,9 +446,76 @@ func (in *ServiceResolverRedirect) toConsul() *capi.ServiceResolverRedirect {
+
+	asJSON, _ := json.Marshal(in)
+	if in.isEmpty() {
+		errs = append(errs, field.Invalid(path, string(asJSON),
```
if `in` is empty won't the `asJSON` also be just the zero values? probably can just leave it out

wilkermichael commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -374,9 +446,76 @@ func (in *ServiceResolverRedirect) toConsul() *capi.ServiceResolverRedirect {
+
+	switch {
+	case in.SamenessGroup != "" && in.ServiceSubset != "":
+		errs = append(errs, field.Invalid(path, string(asJSON),
```
I think this will become pretty long sequence of error strings with repeating the json object each time we append a new error... maybe just scale it back to just the problematic field? or if you don't want to include the json object you could check for an error at the end and add the json object there?

wilkermichael commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -374,9 +446,76 @@ func (in *ServiceResolverRedirect) toConsul() *capi.ServiceResolverRedirect {
+			"redirect.samenessGroup cannot be set with redirect.serviceSubset"))
+	case in.SamenessGroup != "" && in.Partition != "":
+		errs = append(errs, field.Invalid(path, string(asJSON),
+			"redirect.partition cannot be set with redirect.samenessGroup"))
```
I would just check again, but I think `redirect` might already be covered by the path?

wilkermichael commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -513,17 +653,80 @@ func (in *ServiceResolver) validateEnterprise(consulMeta common.ConsulMeta) fiel
 		// error is we have an empty object.
-		return field.Invalid(path, "{}",
-			"service, serviceSubset, namespace, datacenters, policy, and targets cannot all be empty at once")
+		errs = append(errs, field.Invalid(path, "{}",
```
If possible add test coverage for these validations

wilkermichael commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -513,17 +653,80 @@ func (in *ServiceResolver) validateEnterprise(consulMeta common.ConsulMeta) fiel
+	if in.SamenessGroup != "" {
+		switch {
+		case len(in.Datacenters) > 0:
+			errs = append(errs, field.Invalid(path, string(asJSON),
```
same comment as above for including the json object

wilkermichael commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -513,17 +653,80 @@ func (in *ServiceResolver) validateEnterprise(consulMeta common.ConsulMeta) fiel
+		asJSON, _ := json.Marshal(target)
+		switch {
+		case target.Peer != "" && target.ServiceSubset != "":
+			errs = append(errs, field.Invalid(path.Child("targets").Index(i), string(asJSON),
```
same comment about the json object

malizz replied:
resolving this convo: removed validation on subsets

wilkermichael reviewed (changes requested):
@malizz I realised after doing the review that most of the validation logic is duplicated from the Consul repo. I don't think we want to duplicate that validation in Consul-K8s... if there's an error creating a service resolver, that error should be serviced up by the CRD if Consul fails validating it. I worry about the maintainability of duplicating all this validation code.

I think we should probably remove it, and maybe you can do a quick test to verify how the Consul errors get surfaced up by the Consul validation? I'm also not sure if other validation logic added in the past couple PRs was also duplicated? We might want to remove those checks from Consul-k8s as well in a separate PR. 

Let me know if you would like to discuss

malizz commented:
> @malizz I realised after doing the review that most of the validation logic is duplicated from the Consul repo. I don't think we want to duplicate that validation in Consul-K8s... if there's an error creating a service resolver, that error should be serviced up by the CRD if Consul fails validating it. I worry about the maintainability of duplicating all this validation code.
> 
> I think we should probably remove it, and maybe you can do a quick test to verify how the Consul errors get surfaced up by the Consul validation? I'm also not sure if other validation logic added in the past couple PRs was also duplicated? We might want to remove those checks from Consul-k8s as well in a separate PR.
> 
> Let me know if you would like to discuss

Michael and I discussed this offline and we decided that the validations that will need more dynamic context will be performed by Consul core and therefore are removed from this PR. Other static checks are added to validate configs as early as possible to improve UX. 

@wilkermichael your previous comments must be addressed. I removed hasSubset references, and added test coverage. Thanks.

wilkermichael reviewed (commented):
Thanks for trimming down the validation and fixing the coverage, this looks a lot better. Just a couple comments, then I think we're good to go

malizz commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -522,27 +485,35 @@ func (in ServiceResolverFailoverMap) toConsul() map[string]capi.ServiceResolverF
 	m := make(map[string]capi.ServiceResolverFailover)
 	for k, v := range in {
-		m[k] = v.toConsul()
+		m[k] = *v.toConsul()
```
Consul Core checks if redirect or failover values are empty and if so it errors out, also both failover and redirect can't be set at the same time. I think it's better to return a nil rather than an empty struct here. But I will add a line to check for nil values. Thanks for catching this.

malizz replied:
resolving this convo: addressed this comment.

malizz commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -322,14 +329,46 @@ func (in *ServiceResolver) Validate(consulMeta common.ConsulMeta) error {
+		for i, target := range f.Targets {
+			if target.ServiceSubset != "" && (target.Service == "" || target.Service == in.Name) {
+				if !in.Spec.Subsets.hasSubset(target.ServiceSubset) {
+					errs = append(errs, field.Invalid(path.Child("targets").Index(i), "", "serviceSubset is not a valid subset of this service this service resolver"))
```
resolving this convo: removed validation on subsets

malizz commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -322,14 +329,46 @@ func (in *ServiceResolver) Validate(consulMeta common.ConsulMeta) error {
+			(f.Service == "" || f.Service == in.Name) {
+			if !in.Spec.Subsets.hasSubset(f.ServiceSubset) {
+				errs = append(errs, field.Invalid(path, f.ServiceSubset,
+					"failover.serviceSubset is not a valid subset of this service resolver"))
```
resolving this convo: removed validation on subsets

malizz commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -356,6 +395,39 @@ func (in ServiceResolverSubsetMap) toConsul() map[string]capi.ServiceResolverSub
+			errs = append(errs, field.Invalid(indexPath, name, "subset name must begin or end with lower case alphanumeric characters, and contain lower case alphanumeric characters or '-' in between"))
+		}
+		if subset.Filter != "" {
+			if _, err := bexpr.CreateEvaluator(subset.Filter, nil); err != nil {
```
resolving this convo: removed validation on subsets

malizz commented on control-plane/api/v1alpha1/serviceresolver_types.go:
```diff
@@ -322,14 +329,46 @@ func (in *ServiceResolver) Validate(consulMeta common.ConsulMeta) error {
+		(in.Spec.Redirect.Service == "" || in.Spec.Redirect.Service == in.Name) {
+		if !in.Spec.Subsets.hasSubset(in.Spec.Redirect.ServiceSubset) {
+			errs = append(errs, field.Invalid(path, in.Spec.Redirect.ServiceSubset,
+				"redirect.serviceSubset is not a valid subset of this service resolver"))
```
resolving this convo: removed validation on subsets

wilkermichael reviewed (approved):
Looks good, thanks Mali!
